Parse manifest directory

dweepgogia opened the pull request.

dweepgogia committed:
[cli] change from verify to manifest command

- To avoid redundancy while using the cli (operator-verify manifest)
  instead of (operator-verify verify).

dweepgogia committed:
[lib] add support for handling manifest

- To verify the operator manifest.

dweepgogia committed:
[lib] csv validator for example annotations

- To compare example annotations to the provided APIs list
  and report appropriate errors.

dweepgogia committed:
[lib] parse operator manifest directory

- To check for the manifest format and validate the manifest
  and individual operator files.

dweepgogia committed:
[lib] add crd validator

- To support validating the CRDs present in the manifest.

estroz commented on cmd/verify.go:
```diff
@@ -28,8 +28,26 @@ func verifyFunc(cmd *cobra.Command, args []string) {
 	// TODO: return a pointer instead
-	if err := validate.ValidateCSVManifest(yamlFileName); err != (validator.Error{}) {
-		fmt.Println(err)
+	if manifestResultList != nil {
```
Always check if `len(map or slice) ==/!= 0` instead of `==/!= nil`, because `[]type{}` is not `nil` but has a zero length. There are one or two more lines elsewhere that should be changed to the former.

dweepgogia replied:
Good to learn that, thanks!

estroz commented on pkg/validate/crd.go:
```diff
@@ -0,0 +1,66 @@
+func (v *CRDValidator) Validate() (results []validator.ManifestResult) {
+	for _, crd := range v.crds {
+		// Contains error logs for all missing optional and mandatory fields.
+		//result := csvInspect(crd)
```
Remove this line.

estroz commented on pkg/validate/crd.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func crdInspect(crd v1beta1.CustomResourceDefinition) (manifestResult validator.ManifestResult) {
+	scheme := runtime.NewScheme()
```
You can initialize this scheme once in `Validate` and pass it to `crdInspect`.

estroz commented on pkg/validate/csv.go:
```diff
@@ -163,3 +167,94 @@ func isEmptyValue(v reflect.Value) bool {
+	var ok bool
+	annotations := csv.ObjectMeta.GetAnnotations()
+	// Return right away if no examples annotations are found.
+	if annotations == nil {
```
Check length

estroz commented on pkg/validate/csv.go:
```diff
@@ -163,3 +167,94 @@ func isEmptyValue(v reflect.Value) bool {
+		manifestResult.Warnings = append(manifestResult.Warnings, validator.InvalidOperation(fmt.Sprintf("Warning: example annotations not found for %s", csv.GetName()), csv.GetName()))
+		return
+	}
+	// Expect either `alm-examples` or `old.examples` but not both
```
If you cannot use both simultaneously, there should be an error returned if they are.

dweepgogia replied:
I'm not sure if that should raise an error. I can return a warning instead if mentioning both would not affect the operator.

estroz replied:
Either return a warning or change the comment.

dweepgogia replied:
Sure.

gallettilance commented on pkg/validate/crd.go:
```diff
@@ -0,0 +1,66 @@
+		er := validator.Error{Type: validator.ErrorType(err.Type), Field: err.Field, BadValue: err.BadValue, Detail: err.Detail}
+		manifestResult.Errors = append(manifestResult.Errors, er)
+	}
+	return
```
Did you forget to return the `ManifestResult`?

dweepgogia replied:
I'm using golang's `Named Return` concept. See https://tour.golang.org/basics/7.

gallettilance replied:
Sweet! Didn't know about this - pretty cool

gallettilance commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), "kind") {
+			splite := strings.Split(scanner.Text(), " ")
+			if len(splite) < 2 {
```
can you put a comment here to explain why you are gating on this specific condition?

estroz commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), "kind") {
```
Better to do something like what we do in the SDK:

```go
// GetTypeMetaFromBytes gets the type and object metadata from b. b is assumed
// to be a single Kubernetes resource manifest.
func GetTypeMetaFromBytes(b []byte) (t metav1.TypeMeta, err error) {
	u := unstructured.Unstructured{}
	r := bytes.NewReader(b)
	dec := yaml.NewYAMLOrJSONDecoder(r, 8)
	// There is only one YAML doc if there are no more bytes to be read or EOF
	// is hit.
	if err := dec.Decode(&u); err == nil && r.Len() != 0 {
		return t, errors.New("error getting TypeMeta from bytes: more than one manifest in b")
	} else if err != nil && err != io.EOF {
		return t, errors.Wrap(err, "error getting TypeMeta from bytes")
	}
	return metav1.TypeMeta{
		APIVersion: u.GetAPIVersion(),
		Kind:       u.GetKind(),
	}, nil
}
```

So you know you're
1. getting a k8s object
2. not accidentally getting another line containing "kind" that is not a YAML field.

Then you can access kind like `typeMeta.Kind`.

dweepgogia replied:
How about I use `yaml.UnmarshalStrict` for identifying `CSVs` and `CRDs` the way I use it for `package` files?

estroz commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+	isManifestResultNameSet := false
+
+	// parse manifest directory structure
+	_ = filepath.Walk(manifestDirectory, func(path string, f os.FileInfo, err error) error {
```
Best practice is to never throw away errors. You should either add them to the result or (better) log them.

dweepgogia replied:
It does look like that I'm throwing away the error but I am not. All the `return` statements in the `walk` function are returning `nil`. Instead, I'm appending the errors in `manifestResult` object.

estroz replied:
You're not checking the `err` argument, and overwriting it [here](https://github.com/dweepgogia/new-manifest-verification/blob/c5f69a5c86f1832146e69df342b6d659baba01bf/pkg/validate/parser.go#L86). Better to check and ignore it.

estroz commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+
+		// set manifest name
+		if !isManifestResultNameSet {
+			splitPath := strings.Split(path, "/")
```
Use `filepath.Base(path)`

gallettilance commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+	isManifestResultNameSet := false
+
+	// parse manifest directory structure
+	_ = filepath.Walk(manifestDirectory, func(path string, f os.FileInfo, err error) error {
```
this function could be shorter by adding a few helpers.

estroz commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+		}
+		// create a manifest bundle for each version in the manifest
+		if f.IsDir() && path != manifestDirectory {
+			if _, ok := manifest.Bundle[path]; !ok {
```
If `manifest.Bundle` hasn't been initialized this will panic. I suggest doing that above `filepath.Walk()`

gallettilance commented on pkg/validate/csv.go:
```diff
@@ -7,13 +7,15 @@ import (
 
 	"github.com/dweepgogia/new-manifest-verification/pkg/validate/validator"
-	"github.com/pkg/errors"
+	// "github.com/pkg/errors"
```
if this is supposed to be deleted, please delete

estroz commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+				manifest.Bundle[path] = bundle
+			}
+		} else if !f.IsDir() {
+			if strings.Contains(f.Name(), "package") {
```
Better to marshal this into a [`PackageManifest`](https://github.com/operator-framework/operator-registry/blob/f5ee2f9/pkg/registry/types.go#L27) and if there's no error you know its of the correct type.

dweepgogia replied:
[json.Unmarshal()](https://golang.org/pkg/encoding/json/#Unmarshal) does not return an error if the values in the source do not correspond to the values in the target - it just ignores them.

estroz replied:
You can use [`yaml.UnmarshalStrict`](https://godoc.org/sigs.k8s.io/yaml#UnmarshalStrict).

dweepgogia replied:
That's not returning an error either.

dweepgogia replied:
I was using a different `yaml` package. Working now. Thanks!

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -13,38 +13,41 @@ import (
 
 // ValidateCSVManifest takes in name of the yaml file to be validated, reads
 // it, and calls the unmarshal function on rawYaml.
```
Either this comment doesn't make much sense for this function or the name of the function is wrong. Why is CSV in the name is this validates any yaml file?

estroz commented on pkg/validate/parser.go:
```diff
@@ -0,0 +1,121 @@
+				manifestResult.Errors = append(manifestResult.Errors, validator.InvalidManifestStructure(err.Error()))
+				return nil
+			}
+			directoryName := strings.TrimSuffix(path, f.Name())
```
Use `filepath.Dir(path)` here.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -13,38 +13,41 @@ import (
 	if err != nil {
-		return validator.IOError(fmt.Sprintf("Error in reading %s file:   #%s ", yamlFileName, err), yamlFileName)
+		manifestResult.Errors = append(manifestResult.Errors, validator.IOError(fmt.Sprintf("Error in reading %s file:   #%s ", yamlFileName, err), yamlFileName))
+		return
```
why are you not returning `manifestResult` here?

dweepgogia replied:
Same as above

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -13,38 +13,41 @@ import (
 	if err != nil {
-		return validator.InvalidParse(fmt.Sprintf("Error unmarshalling YAML to OLM's csv type for %s file:  #%s ", yamlFileName, err), yamlFileName)
+		manifestResult.Errors = append(manifestResult.Errors, validator.InvalidParse(fmt.Sprintf("Error unmarshalling YAML to OLM's csv type for %s file:  #%s ", yamlFileName, err), yamlFileName))
+		return
```
here too

dweepgogia replied:
Same as above.

estroz commented on pkg/validate/serialize.go:
```diff
@@ -13,38 +13,41 @@ import (
-	fmt.Println("Running", v.Name())
+	fmt.Println("\nRunning", v.Name())
 	for _, errorLog := range v.Validate() {
 		fmt.Println("Validating CSV", errorLog.Name)
```
```suggestion
		fmt.Printf("Validating CSV %s\n\n", errorLog.Name)
```
and remove `fmt.Println()` below.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -13,38 +13,41 @@ import (
 	if err := v.AddObjects(csv); err != (validator.Error{}) {
-		return err // TODO: update when 'AddObjects' returns an actual error.
+		manifestResult.Errors = append(manifestResult.Errors, err)
+		return // TODO: update when 'AddObjects' returns an actual error.
```
also here

dweepgogia replied:
Same as above.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -13,38 +13,41 @@ import (
-	fmt.Printf("%s is verified.\n", yamlFileName)
-	return validator.Error{}
+	fmt.Printf("\n%s is verified.\n", yamlFileName)
+	return
```
and here

dweepgogia replied:
Same as above.

gallettilance commented on pkg/validate/serialize.go:
```diff
@@ -72,11 +75,23 @@ func unmarshal(rawYAML []byte) (olm.ClusterServiceVersion, error) {
+		return manifestResultList
+	}
+	for _, bundle := range manifest.Bundle {
+		manifestResultList = append(manifestResultList, ValidateCSVManifest(bundle.CSV))
```
so you are only validating the CSVs in this commit?

gallettilance replied:
can we try to validate the other stuff too? What's blocking?

estroz reviewed (changes requested):
First pass: looks ok overall; the bundle parser could probably use some refactoring to reduce complexity.

gallettilance reviewed (commented):
Great work! A few nit picks, questions, requests - let me know

estroz commented:
We should explore using https://godoc.org/golang.org/x/xerrors for returned errors somehow instead of custom return types. That package (coming to the std lib in Go 1.13) provides a nice interface for creating, type checking, and unwrapping errors. Outside of this PR's scope though.

dweepgogia commented:
> We should explore using https://godoc.org/golang.org/x/xerrors for returned errors somehow instead of custom return types. That package (coming to the std lib in Go 1.13) provides a nice interface for creating, type checking, and unwrapping errors. Outside of this PR's scope though.

We decided to go with custom error reporting, especially the way we have defined it, because it's compatible with the way Kubernetes reports errors for all the validations they're performing. You can have a look at their definition here: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/field/errors.go#L31. 
If in the future, we move to the upstream this library would support that transition, and there should not be many changes required.

dweepgogia committed:
[lib] add package validator

- To support validating package yaml present in the manifest.

dweepgogia committed:
[lib] add bundle validator

- To perform validation across manifest files.

dweepgogia commented:
Updated the PR with suggested changes and added two additional commits for `package` and `bundle` validators. Please review when you get a chance.

Note: will connect the `CRD` and `package` validators in upcoming PRs.

estroz commented:
@dweepgogia sounds good, keep the current error type as-is.

dweepgogia commented:
Thanks for reviewing @estroz and @galletti94

estroz reviewed (changes requested):
While the functional ordering here works, its a little rough. I'd rather not see print statements in functions called by non-top level functions and returned values thrown away, ex. [here](https://github.com/dweepgogia/new-manifest-verification/pull/10/commits/79a4c672b5e609125bd5872c941751bbc966cb0c#diff-bdc50831c9bdaed93469e0e72394e9e0R184). I'd like to see a flow like:

1. Parse files/directory
    * You might want to add type information to each file, ex. "this file is a CSV", so you can validate them by type
1. Run validator helpers on each file parsed
1. Check results and print errors

This flow is cleaner, more logical, and uses function return values as another library user would.

dweepgogia committed:
[lib] add two methods to validator interface

- To improve the design.

Commit:
Merge branch 'master' into parse-manifest-directory
